core/daemon/src/main/java/org/opennms/netmgt/vmmgr/Manager.java

```diff
@@ -54,61 +55,54 @@
 
 /**
  * <p>
- * The Manager is reponsible for launching/starting all services in the VM
- * that it is started for. The Manager operates in two modes, normal and
- * server
+ * The Manager is responsible for managing running services in the VM
+ * that it is started for.
  * </p>
  * <p>
- * normal mode: In the normal mode, the Manager starts all services configured
- * for its VM in the service-configuration.xml and starts listening for
- * control events on the 'control-broadcast' JMS topic for stop control
- * messages for itself
+ * The Starter starts all services configured for its VM in the
+ * service-configuration.xml file and the manager exposes an
+ * MBean over JMX that is used to get service status and stop services.
  * </p>
- * <p>
- * server mode: In the server mode, the Manager starts up and listens on the
- * 'control-broadcast' JMS topic for 'start' control messages for services in
- * its VM and a stop control messge for itself. When a start for a service is
- * received, it launches only that service and sends a successful 'running' or
- * an 'error' response to the Controller
- * </p>
- * <p>
- * <strong>Note: </strong>The Manager is NOT intelligent - if it receives a
- * stop control event, it will exit - does not check to see if the services
- * its started are all stopped
- * <p>
  *
  * @author <a href="mailto:weave@oculan.com">Brian Weaver</a>
  * @author <a href="mailto:sowmya@opennms.org">Sowmya Nataraj</a>
  * @author <a href="http://www.opennms.org">OpenNMS.org</a>
  */
 public class Manager implements ManagerMBean {
 
-	private static final Logger LOG = LoggerFactory.getLogger(Manager.class);
+    private static final Logger LOG = LoggerFactory.getLogger(Manager.class);
 
     /**
      * The log4j category used to log debug messages and statements.
      */
     private static final String LOG4J_CATEGORY = "manager";
     private static final String m_osName = System.getProperty("os.name") == null? "" : System.getProperty("os.name").toLowerCase();
-    private static long startTime = System.currentTimeMillis();
-    private static AtomicBoolean stopInitiated = new AtomicBoolean(false);
+    private static final long startTime = System.currentTimeMillis();
+    private static final AtomicBoolean stopInitiated = new AtomicBoolean(false);
+    private static final AtomicInteger exitCode = new AtomicInteger(0);
 
 
     /**
      *  Register shutdown hook to handle SIGTERM signal for the process.
      */
     public void init() {
-        Runtime.getRuntime().addShutdownHook(new Thread(() -> stop()));
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> { LOG.info("got sigterm, stopping"); stop(); }));
     }
+
     /**
      * <p>stop</p>
      */
     @Override
     public synchronized void stop() {
+        stop(0);
+    }
+
+    public synchronized void stop(int processExitCode) {
         if (stopInitiated.get()) {
```

Comment:
I wonder if you should check the existing error code here, and set it if `processExitCode` is non-zero, but the current value is zero?

IE, something calls stop, but then something else calls it with an error code, I would think you'd want to propagate the error code.

Reply:
Good suggestion! And it made the change simpler, too. :) This is the updated commit: https://github.com/OpenNMS/opennms/pull/5769/commits/2dad56d323ecf16722cf83e4f65baa814b9783aa